Modify access specifier to protected or public for the scope of processMessage() member function

hyunseok-yang opened the pull request:
- pose_display, pose_with_covariance_display and relative_humidity_display

hyunseok-yang committed:
Modify access specifier to protected or public for the scope of processMessage() member function
- pose_display, pose_with_covariance_display
- marker_display, marker_array_display
- point_cloud_common
- relative_humidity_display, temperature_display, fluid_pressure_display, illuminance_display

Signed-off-by: (=YG=) Hyunseok Yang <hyunseok7.yang@lge.com>

clalancette commented on rviz_default_plugins/include/rviz_default_plugins/displays/relative_humidity/relative_humidity_display.hpp:
```diff
@@ -56,7 +56,7 @@ class RVIZ_DEFAULT_PLUGINS_PUBLIC RelativeHumidityDisplay
   ~RelativeHumidityDisplay() override;
 
-private:
+protected:
```
By switching this to `protected`, you are also making `setInitialValues` and `hideUnneededProperties` available to sub-classes.  Do we really need that?

hyunseok-yang replied:
Hi,
well I just modified those for consistency of display pluging template.
I didn't mean to make visible to sub-classes but I think that's not a big deal.

And these functions are also virtual function of parent's so that another sub-classese can override that functions.
(although not necessary to me frankly :) )

So, unless it is public, it would be NO problem.

```c++
  void setInitialValues() override;
  void hideUnneededProperties() override;
```

clalancette replied:
> And these functions are also virtual function of parent's so that another sub-classese can override that functions.

Right, but these are the derived classes, not the base class.

I guess I'll say that in every single other class that derives from `PointCloudScalarDisplay`, these two methods are private.  So I'd prefer to keep that consistent here, and leave these methods private.

hyunseok-yang replied:
To me, all I need is override `processMessage()` function.  So it's fine to keep two methods as a private functions.

BTW, is it correct that you mentioned?

https://github.com/ros2/rviz/blob/bc93b3488e9b204b1e0e9df2455e1c66c552cb41/rviz_default_plugins/include/rviz_default_plugins/displays/pointcloud/point_cloud_scalar_display.hpp#L97

Cause you said those are 'private' but it written as a protected :)

hyunseok-yang replied:
> > And these functions are also virtual function of parent's so that another sub-classese can override that functions.
> 
> Right, but these are the derived classes, not the base class.
> 
> I guess I'll say that in every single other class that derives from `PointCloudScalarDisplay`, these two methods are private. So I'd prefer to keep that consistent here, and leave these methods private.

To me, all I need is override `processMessage()` function.  So it's fine to keep two methods as a private functions.

BTW, is it correct that you mentioned?

https://github.com/ros2/rviz/blob/bc93b3488e9b204b1e0e9df2455e1c66c552cb41/rviz_default_plugins/include/rviz_default_plugins/displays/pointcloud/point_cloud_scalar_display.hpp#L97

Cause you said those are 'private' but it written as a protected :)

clalancette replied:
> Cause you said those are 'private' but it written as a protected :)

I meant `private`.  `setInitialValues` and `hideUnneededProperties` are private in every class that derives from `PointCloudScalarDisplay`, as far as I can tell.

hyunseok-yang replied:
> > Cause you said those are 'private' but it written as a protected :)
> 
> I meant `private`. `setInitialValues` and `hideUnneededProperties` are private in every class that derives from `PointCloudScalarDisplay`, as far as I can tell.

I see, now I get it. I didn't check another class such as `FluidPressureDisplay` which is derivied from PointCloudScalarDisplay`

hyunseok-yang replied:
I will modify it to 'private' for consistency in terms of what you mentioned above.

And will modify more plugins for `processMessage()` function.

hyunseok-yang commented:
@ahcorde @clalancette  I rebased the branch on latest rolling. And fixed all related to access specifier for 'processMessage()' function.

@clalancette is it make sense? :)

clalancette commented on rviz_default_plugins/include/rviz_default_plugins/displays/fluid_pressure/fluid_pressure_display.hpp:
```diff
@@ -47,7 +47,7 @@ namespace displays
  */
 
-class RVIZ_DEFAULT_PLUGINS_PUBLIC FluidPressureDisplay
+class RVIZ_DEFAULT_PLUGINS_PUBLIC z
```
This is a typo.

```suggestion
class RVIZ_DEFAULT_PLUGINS_PUBLIC FluidPressureDisplay
```

hyunseok-yang committed:
Update rviz_default_plugins/include/rviz_default_plugins/displays/fluid_pressure/fluid_pressure_display.hpp

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

hyunseok-yang reviewed (commented):
d'oh
thanks

clalancette reviewed (approved):
Thanks for iterating.  This looks good to me; I'll run CI on it next.

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=18757)](http://ci.ros2.org/job/ci_linux/18757/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=13282)](http://ci.ros2.org/job/ci_linux-aarch64/13282/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=19500)](http://ci.ros2.org/job/ci_windows/19500/)

hyunseok-yang commented:
@clalancette 
Hi, Could you let me know when this changes can merged into humble branch?
I just wonder how the CI/CID process works :)

clalancette commented:
> Hi, Could you let me know when this changes can merged into humble branch?

We have to consider further before backporting to humble.  In particular, we have to be sure these changes don't change ABI.  I don't *think* that they do, but I need to consider it further.

hyunseok-yang commented:
> 

OK. Thanks for your comment.
Have a good day

clalancette commented:
Unfortunately, looking at this closer, it looks like this has the potential to change ABI.  In particular, see https://community.kde.org/Policies/Binary_Compatibility_Examples#Change_the_access_rights .  So we cannot backport this to `humble` as-is.

That said, there is a workaround mentioned in https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B (search for "Changing the access rights").  In particular, you could make a new `protected` method that called the existing `private` method.  If you'd like to pursue that path, I'm happy to review the resulting PR.

hyunseok-yang commented:
> Unfortunately, looking at this closer, it looks like this has the potential to change ABI. In particular, see https://community.kde.org/Policies/Binary_Compatibility_Examples#Change_the_access_rights . So we cannot backport this to `humble` as-is.
> 
> That said, there is a workaround mentioned in https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B (search for "Changing the access rights"). In particular, you could make a new `protected` method that called the existing `private` method. If you'd like to pursue that path, I'm happy to review the resulting PR.

Thanks for your review.
Understood in terms of 'changing the access rights'.

I'd like to make a new `protected` method which calls existing `private` function. I'm not sure this path makes consistency with other class. 
And then, most important things is `processMessage()` function is called by somewhere else as a binding function. So it's not easy to implement what I expected.

I think workaround solution for just humble is not a way of good design.


Is there any other option to backport into humble? :(

clalancette commented:
> Is there any other option to backport into humble? :(

Unfortunately no.  We have to keep ABI in humble.
